compiler-rt/lib/profile/InstrProfilingFile.c

```diff
@@ -894,18 +894,27 @@ static void parseAndSetFilename(const char *FilenamePat,
                                 ProfileNameSpecifier PNS,
                                 unsigned CopyFilenamePat) {
 
-  const char *OldFilenamePat = lprofCurFilename.FilenamePat;
+  char *OldFilenamePat = lprofCurFilename.FilenamePat
+                             ? strdup(lprofCurFilename.FilenamePat)
+                             : NULL;
   ProfileNameSpecifier OldPNS = lprofCurFilename.PNS;
 
   /* The old profile name specifier takes precedence over the old one. */
-  if (PNS < OldPNS)
+  if (PNS < OldPNS) {
+    if (OldFilenamePat) {
+      free(OldFilenamePat);
+      OldFilenamePat = NULL;
```

Comment:
I don't quite see the need for setting the variable to NULL right before returning - it's a local variable after all? Plus you don't need to check whether it's non-null before invoking `free()`, it's ok to call `free(NULL)`. So it should be enough with just one line with `free(OldFilenamePat);` in each codepath.